server/src/test/java/io/crate/integrationtests/CopyFromFailFastITest.java

```diff
@@ -0,0 +1,276 @@
+        execute("set global overload_protection.dml.queue_size = 2");
+
+        execute("CREATE TABLE doc.t (a INT PRIMARY KEY, b INT) CLUSTERED INTO 2 SHARDS WITH (number_of_replicas=0)");
+
+        execute("SELECT node['name'] FROM sys.shards WHERE table_name='t' ORDER BY id");
+        var nodeNameOfShard0 = (String) response.rows()[0][0];
+
+        var clientProvider = new SQLTransportExecutor.ClientProvider() {
+
+            @Override
+            public Client client() {
+                return internalCluster().client(nodeNameOfShard0);
+            }
+
+            @Override
+            public String pgUrl() {
+                return null;
+            }
+
+            @Override
+            public SQLOperations sqlOperations() {
+                return internalCluster().getInstance(SQLOperations.class, nodeNameOfShard0);
+            }
+        };
+        var handlerNodeExecutor = new SQLTransportExecutor(clientProvider);
+
+        Path tmpDir = newTempDir(LifecycleScope.TEST);
+        Path target = Files.createDirectories(tmpDir.resolve("target"));
+        int numDocs = 100;
+
+        var indexMetadata = clusterService().state().getMetadata().index("t");
+        List<String> rows = new ArrayList<>();
+        for (int i = 0; i < numDocs; i++) {
+            String line;
+            // Ensure the failing doc will fail on shard id 0
+            if (i >= 20 && OperationRouting.generateShardId(indexMetadata, Integer.toString(i), null) == 0) {
+                line = "{\"a\":" + i + ", \"b\":\"fail here\"}";
+            } else {
+                line = "{\"a\":" + i + ", \"b\":" + i + "}";
+            }
+            rows.add(line);
+        }
+
+        tmpFileWithLines(rows, "data1.json", target);
+        assertExpectedLogMessages(
+            () -> assertThrowsMatches(
+                () -> handlerNodeExecutor.exec(
+                    "COPY doc.t FROM ? WITH (bulk_size = 1, fail_fast = true, shared= true)", // fail_fast = true
+                    new Object[]{target.toUri() + "*"}),
+                JobKilledException.class,
+                "failed to parse field [b] of type [integer]"
+            ),
+            new MockLogAppender.PatternSeenEventExcpectation(
+                "assert failure on node=" + nodeNameOfShard0,
+                "io.crate.execution.dml.upsert.TransportShardUpsertAction",
+                Level.DEBUG,
+                "Failed to execute upsert on nodeName="+nodeNameOfShard0 + ".*")
+        );
+
+        execute("SELECT COUNT(*) FROM doc.t");
+        assertThat((long) response.rows()[0][0], lessThan(50L));
+    }
+
+    @Test
+    public void test_copy_from_fail_fast_without_return_summary_without_write_errors() throws IOException {
+        // fail_fast relies on return-summary logic in order to generate meaningful exception messages if fails.
+        // If passes, from the collected upsertResults, only the successRowCounts is used.
```

Comment:
This comment still sounds to me as the new `fail-fast` logic does not change any logic if everything succeeds. Such I don't think this test case covers anything new. But also ok to keep it.